File: hash/go/imagehash/internal/testkit/predecoded.go
package testkit

import (
	"encoding/binary"
	"image"
	"os"
	"testing"
)

// LoadPreDecodedFromRoot reads spec/decoded/<name>.rgb.bin and returns it as
// a *image.NRGBA with canonical RGB layout (A=255 in every pixel).
func LoadPreDecodedFromRoot(t *testing.T, name string) image.Image {
	t.Helper()
	return loadPreDecoded(t, DirRoot(), name)
}

// LoadPreDecodedFromInternal is the same, for internal/* tests.
func LoadPreDecodedFromInternal(t *testing.T, name string) image.Image {
	t.Helper()
	return loadPreDecoded(t, DirInternal(), name)
}

func loadPreDecoded(t *testing.T, specDir, name string) image.Image {
	t.Helper()
	path := specDir + "/decoded/" + name + ".rgb.bin"
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	if len(data) < 8 {
		t.Fatalf("decoded file %s too short: %d bytes", name, len(data))
	}
	w := int(binary.LittleEndian.Uint32(data[0:4]))
	h := int(binary.LittleEndian.Uint32(data[4:8]))
	expected := 8 + w*h*3
	if len(data) != expected {
		t.Fatalf("decoded %s length mismatch: got %d, expected %d", name, len(data), expected)
	}
	img := image.NewNRGBA(image.Rect(0, 0, w, h))
	off := 8
	for y := 0; y < h; y++ {
		for x := 0; x < w; x++ {
			i := img.PixOffset(x, y)
			img.Pix[i+0] = data[off]
			img.Pix[i+1] = data[off+1]
			img.Pix[i+2] = data[off+2]
			img.Pix[i+3] = 255
			off += 3
		}
	}
	return img
}

File: hash/rust/rosetta-image-hash/tests/testkit/predecoded.rs
//! Loads `spec/decoded/<name>.rgb.bin` into a `DynamicImage::ImageRgb8`.

use std::fs;

use image::{DynamicImage, ImageBuffer, Rgb};

use super::spec_path;

pub fn load_predecoded(name: &str) -> DynamicImage {
    let path = format!("{}/decoded/{}.rgb.bin", spec_path::SPEC_DIR, name);
    let data = fs::read(&path).unwrap_or_else(|e| panic!("read {path}: {e}"));
    assert!(data.len() >= 8, "decoded file {name} too short");
    let w = u32::from_le_bytes(data[0..4].try_into().unwrap());
    let h = u32::from_le_bytes(data[4..8].try_into().unwrap());
    let expected = 8 + (w as usize) * (h as usize) * 3;
    assert_eq!(
        data.len(),
        expected,
        "decoded {name} length mismatch: got {}, expected {expected}",
        data.len()
    );
    let buf: ImageBuffer<Rgb<u8>, Vec<u8>> =
        ImageBuffer::from_raw(w, h, data[8..].to_vec()).expect("ImageBuffer construction");
    DynamicImage::ImageRgb8(buf)
}
